rainier-core/src/main/scala/com/stripe/rainier/sampler/LongestBatch.scala

```diff
@@ -0,0 +1,97 @@
+package com.stripe.rainier.sampler
+
+/**
+  * Compute a distribution of number of leapfrog steps required to make
+  * a u-turn using a tuned value of the step size eps
+  * @param l0 the initial number of leapfrog steps
+  * @param eps the tuned leapfrog step size
+  */
+object LongestBatch {
+
+  /**
+    * Calculate the longest-step size until a u-turn
+    */
+  def longestBatch(params: Array[Double],
+                   l0: Int,
+                   eps: Double,
+                   lf: LeapFrog) = {
+
+    val initTheta = lf.variables(params)
+    var out = params
+    var pq = params
+    var l = 0
+    while (LongestBatch.uturnCondition(initTheta,
+                                       lf.variables(pq),
+                                       lf.momentum(pq))) {
+      l += 1
+      val pqPlus = lf.leapfrogs(1, eps, pq)
+      if (l == l0) {
+        out = pqPlus
+        pq = pqPlus
+      } else {
+        pq = pqPlus
+      }
+
+    }
+    (out, l)
+  }
+
+  /**
+    * Perform a single step of the longest batch step algorithm
+    */
+  def longestBatchStep(l0: Int, eps: Double, lf: LeapFrog)(
+      thetal: (Array[Double], Int))(implicit rng: RNG): (Array[Double], Int) = {
+
+    val params = thetal._1
+    lf.initializePs(params)
+    val (tp, l) = longestBatch(params, l0, eps, lf)
+    val prop = if (l < l0) {
+      lf.leapfrogs(l0 - l, eps, tp)
```

Comment:
Once in `LeapFrog`, you should model this off `step`, in terms of copying params into pqBuf and then having `leapfrogs` modify that - rather than returning a new array.

Reply:
In the case where `l < l0`, then `tp` will be the same as `params` (because of how `longestBatch` works. Does that mean you'd want to advance `l0` steps instead of `l0 - l`?

In general, it looks to me like we're trying to do two separate things that are weirdly entangled:
- advance exactly `l0` steps and record the param values at that point
- discover how many steps we need until a u-turn

Is that correct? I wonder if we can separate those two concerns out more cleanly without wasting any computation. It seems like it would be good to, if we can, whereas right now they're both kinda smeared across this method and `longestBatch`.

Reply:
That's exactly what we're trying to do, but without duplicating computations. I'll have a think about how exactly to separate this.